helm: add missing helm option to overwrite auto-detected MTU and missing tunnelPort helm document

vincentmli opened the pull request:
cilium agent has MTU option below and also allow users to set it
in configmap cilium-config as mtu: "1500", but it is missing in helm.

--mtu int Overwrite auto-detected MTU of underlying network

found this option through https://github.com/cilium/cilium/issues/20603

add helm option --set MTU=1500 so user could use it
in helm install when needed.

also add missing helm tunnelPort document

Signed-off-by: Vincent Li <v.li@f5.com>

Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Thanks for contributing!

<!-- Description of change -->

Fixes: #issue-number

```release-note
add helm option configuredMTU to overwrite auto-detected MTU and tunnelPort helm document
```

vincentmli committed:
helm: add missing helm option to overwrite auto-detected MTU

cilium agent has MTU option below and also allow users to set it
in configmap cilium-config as mtu: "1500", but it is missing in helm.

--mtu int Overwrite auto-detected MTU of underlying network

found this option through https://github.com/cilium/cilium/issues/20603

add helm option --set MTU=1500 so user could use it
in helm install when needed.

Signed-off-by: Vincent Li <v.li@f5.com>

gandro reviewed (commented):
> trying to fresh my memory on helm related patch, am I also supposed to add the option in `install/kubernetes/cilium/README.md` ?

Yes, please add it (commented out) to the `values.yaml.tmpl` file, and then run `make -C install/kubernetes` and `make -C Documentation` to properly document it. Thanks!

joestringer commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1443,6 +1443,9 @@ tls:
 tunnel: "vxlan"
 
+# -- Configure the underlying network MTU to overwrite auto-detected MTU.
+configuredMTU: ""
```
All Helm options are configured options, any reason not to just define this as MTU?

```suggestion
MTU: ""
```

(Or we could argue that MTU should be nested under datapath instead of being directly at the top level configuration for Cilium, I don't have a strong opinion either way).

vincentmli replied:
it is a name I steal from inside the code, I thought about name `MTU` but   I knew someone like you  is going to suggest a better name :-).

gandro commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -1440,6 +1440,12 @@ tls:
 tunnel: "vxlan"
 
+# -- Configure the VXLAN and Geneve tunnel port
+tunnelPort: ""
```
I believe defaulting to a string here will break the config map validation, since this is expected to be an integer. Note that you can also leave helm values commented out like this:

https://github.com/cilium/cilium/blob/488dbae2c3392d299e08e310ced73085ff051199/install/kubernetes/cilium/values.yaml#L1923-L1924

Therefore I prefer to have those two value commented out by default (the Helm documentation will still be generated, even if commented out)

Edit: This is also observable in the smoke test (it applies both to MTU and tunnelPort):
https://github.com/cilium/cilium/runs/7529915795?check_suite_focus=true

```
2022-07-26T22:14:20.649231622Z level=fatal msg="Incorrect config-map flag value" error="option mtu: unable to cast \"\" of type string to int64" subsys=config
```

vincentmli replied:
how do I define the type of both helm `MTU` and `tunnelPort`?  `# tunnelPort: 8472` in `cilium/values.yaml` would do it?

vincentmli replied:
ok I am still  not sure  how to set the helm MTU type, here is the change I made and committed locally

```
diff --git a/Documentation/helm-values.rst b/Documentation/helm-values.rst
index e77e617c30..8aa12095c6 100644
--- a/Documentation/helm-values.rst
+++ b/Documentation/helm-values.rst
@@ -9,6 +9,10 @@
      - Description
      - Type
      - Default
+   * - MTU
+     - Configure the underlying network MTU to overwrite auto-detected MTU.
+     - string
+     - ``""``
    * - affinity
      - Affinity for cilium-agent.
      - object

diff --git a/Documentation/spelling_wordlist.txt b/Documentation/spelling_wordlist.txt
index ecd777249d..6db168a00e 100644
--- a/Documentation/spelling_wordlist.txt
+++ b/Documentation/spelling_wordlist.txt
@@ -275,6 +275,7 @@ configMap
 configMapKey
 configmap
 configs
+MTU
 conformant
 conntrack
 conntrackGCInterval

diff --git a/install/kubernetes/cilium/README.md b/install/kubernetes/cilium/README.md
index dc15be45d0..55659cd2d8 100644
--- a/install/kubernetes/cilium/README.md
+++ b/install/kubernetes/cilium/README.md
@@ -137,6 +137,7 @@ contributors across the globe, there is almost always someone available to help.
 | cni.hostConfDirMountPath | string | `"/host/etc/cni/net.d"` | Configure the path to where the CNI configuration directory is mounted inside the agent pod. |
 | cni.install | bool | `true` | Install the CNI configuration and binary files into the filesystem. |
 | cni.logFile | string | `"/var/run/cilium/cilium-cni.log"` | Configure the log file for CNI logging with retention policy of 7 days. Disable CNI file logging by setting this field to empty explicitly. |
+| MTU | int | `1500` | Configure the underlying network MTU to overwrite auto-detected MTU. |
 | containerRuntime | object | `{"integration":"none"}` | Configure container runtime specific integration. |
 | containerRuntime.integration | string | `"none"` | Enables specific integrations for container runtimes. Supported values: - containerd - crio - docker - none - auto (automatically detect the container runtime) |
 | customCalls | object | `{"enabled":false}` | Tail call hooks for custom eBPF programs. |

diff --git a/install/kubernetes/cilium/templates/cilium-configmap.yaml b/install/kubernetes/cilium/templates/cilium-configmap.yaml
index fbe8202d69..0730291a98 100644
--- a/install/kubernetes/cilium/templates/cilium-configmap.yaml
+++ b/install/kubernetes/cilium/templates/cilium-configmap.yaml
@@ -395,6 +395,10 @@ data:
   tunnel-port: "{{ .Values.tunnelPort }}"
 {{- end }}
 
+{{- if hasKey .Values "MTU" }}
+  mtu: "{{ .Values.MTU }}"
+{{- end }}
+
 {{- if .Values.eni.enabled }}
   enable-endpoint-routes: "true"
   auto-create-cilium-node-resource: "true"

diff --git a/install/kubernetes/cilium/values.yaml b/install/kubernetes/cilium/values.yaml
index bbdccc1465..20014150c1 100644
--- a/install/kubernetes/cilium/values.yaml
+++ b/install/kubernetes/cilium/values.yaml
@@ -1443,6 +1443,9 @@ tls:
 #   - geneve
 tunnel: "vxlan"
 
+# # -- Configure the underlying network MTU to overwrite auto-detected MTU.
+#   MTU: 1500
+
 # -- Disable the usage of CiliumEndpoint CRD.
 disableEndpointCRD: "false"
 
diff --git a/install/kubernetes/cilium/values.yaml.tmpl b/install/kubernetes/cilium/values.yaml.tmpl
index 8c1e7af740..ffc7c9b9ab 100644
--- a/install/kubernetes/cilium/values.yaml.tmpl
+++ b/install/kubernetes/cilium/values.yaml.tmpl
@@ -1440,6 +1440,9 @@ tls:
 #   - geneve
 tunnel: "vxlan"
 
+# # -- Configure the underlying network MTU to overwrite auto-detected MTU.
+#   MTU: 1500
+
 # -- Disable the usage of CiliumEndpoint CRD.
 disableEndpointCRD: "false"
 
```

then I ran `make -C install/kubernetes cilium/values.yaml`, no changes made on files, good so far
then I ran `make -C Documentation`, this creates `Documentation/helm-values.rst` with 

```
[root@cilium-dev cilium]# git diff
diff --git a/Documentation/helm-values.rst b/Documentation/helm-values.rst
index 8aa12095c6..e77e617c30 100644
--- a/Documentation/helm-values.rst
+++ b/Documentation/helm-values.rst
@@ -9,10 +9,6 @@
      - Description
      - Type
      - Default
-   * - MTU
-     - Configure the underlying network MTU to overwrite auto-detected MTU.
-     - string
-     - ``""``
    * - affinity
      - Affinity for cilium-agent.
      - object
```
now MTU is removed from `/Documentation/helm-values.rst`, MTU is not set to `int` @gandro what I did wrong, or what is correct steps to make proper changes?

gandro replied:
@vincentmli This issue is still unresolved. While my comment was pointing out the tunnelPort, the issue also exists for the MTU variable.

vincentmli replied:
@gandro @joestringer,  sorry for the confusion, do you see the fresh PR update? I removed the tunnelPort commit for now and just kept the MTU commit, I want to make sure I understand it correct what files needs to be changed and changed properly, then run `make -C install/kubernetes`, and `make -C Documentation`.  here is my understanding and steps so far:

1 make changes in 
```
install/kubernetes/cilium/templates/cilium-configmap.yaml
install/kubernetes/cilium/values.yaml.tmpl
install/kubernetes/cilium/values.yaml
install/kubernetes/cilium/README.md ( set the `int` type here in README.md)
```

2 run 
```
make -C install/kubernetes cilium/values.yaml
make -C Documentation <= this creates Documentation/helm-values.rst with user visible documents with correct type

```
3 commit the changes in `Documentation/helm-values.rst` locally and `git rebase -i HEAD~2`  to consolidate into one commit and push to update the PR.

apparently I am missing something

 step 1:  where to set the type correctly
 step 2: make sure `make -C Documentation` generates the correct `Documentation/helm-values.rst`

vincentmli replied:
> I believe defaulting to a string here will break the config map validation, since this is expected to be an integer. Note that you can also leave helm values commented out like this:
> 
> https://github.com/cilium/cilium/blob/488dbae2c3392d299e08e310ced73085ff051199/install/kubernetes/cilium/values.yaml#L1923-L1924
> 
> Therefore I prefer to have those two value commented out by default (the Helm documentation will still be generated, even if commented out)
> 

Hi @gandro I did further test, it looks to me if I comment it out as you suggested like below in `install/kubernetes/cilium/values.yaml.tmpl` 
```
# # -- Configure VXLAN and Geneve tunnel port.
#   tunnelPort: 0

```
`make -C Documentation` will not generate new `Documentation/helm-values.rst`, are you sure comment out still allow new helm documentation generation ?

gandro commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1443,6 +1443,9 @@ tls:
 tunnel: "vxlan"
 
+# # -- Configure the underlying network MTU to overwrite auto-detected MTU.
+#   MTU: 1500
```
```suggestion
#   MTU: 0
```

One last nit: Maybe let's use `0` (autodetect) as the default here. Still commented out though. Thanks.

vincentmli replied:
ok, I will do that, just you know I manually edited `Documentation/helm-values.rst` with the `int` type ( I know I am not suppose to do this way since it is auto generated) and pushed the changes to this PR.

vincentmli committed:
helm: add missing helm tunnelPort option in helm document

commit '030077272e9' added tunnelPort helm option, but missing
in helm documentation

Signed-off-by: Vincent Li <v.li@f5.com>

gandro reviewed (approved):
Thanks @vincentmli
